C2LC-588: Reorganise "movements" and "controls".

the-t-in-rtf opened the pull request:
See [C2LC-588](https://issues.fluidproject.org/browse/C2LC-588) for details.

chosww commented on src/App.scss:
```diff
@@ -164,28 +164,47 @@ body {
-.App__command-palette-commands {
-    width: 6rem;
+.App__command-palette-section-heading {
+    border-bottom: 1px solid #818A98;
```
This border is only applied underneath the text, I think it would be cleaner if we wrap the headings in a container and apply border-bottom on the container.

chosww replied:
We can have a container for the headings("Movements" and "Controls") that has margin top and bottom to make them more centered. Right now they are not vertically centered; Movements have more space at the top and Controls have more space at the bottom. After adding a container for them, we need to remove some padding from existing containers.

chosww replied:
We can give some margin-left in the container I suggested to push the text little bit towards right.

chosww commented on src/App.scss:
```diff
@@ -164,28 +164,47 @@ body {
     overflow-y: auto;
     grid-row-start: 3;
     grid-row-end: 4;
     padding-top: 1rem;
```
After adding containers for the headings, we can remove this padding.

chosww commented on src/App.scss:
```diff
@@ -164,28 +164,47 @@ body {
+    color: #F1AE5B;
+    font-size: .9rem;
+    text-align: left;
+    padding-bottom: .6rem;
```
We can remove this padding-bottom and padding-left from the heading h3 after wrap them in a container.

chosww replied:
We also need to apply margin-top and margin-bottom 0 to the heading h3, because default h3 applies some margin-top and margin-bottom.

chosww commented on src/App.scss:
```diff
@@ -164,28 +164,47 @@ body {
+}
+
+.App__command-palette-commands {
+    width: 3rem;
```
There's some extra spaces to the right of the buttons when we set fixed width of 3rem, as each blocks have width of 2.5rem. That causes layout to be not horizontally centered; we can remove this width.

chosww commented:
<img width="735" alt="Screen Shot 2022-02-03 at 1 46 00 PM" src="https://user-images.githubusercontent.com/22668696/152408794-8398d8b1-f68e-42f5-9365-bc36be179ee5.png">

When I look at the Figma design, I see that different sections of the actions panel is actually in separate containers on top of a darker container. I think right approach would be to make App__command-palette-command-container to have darker background and make separate containers for movements and controls section with round border, then apply box-shadow bottom on each of them.

the-t-in-rtf commented:
@chosww, I tried to incorporate your feedback.  The sections now have their own container, and the heading has a wrapper as well.  I'm still not 100% happy with the shortness of the controls block versus the movements, but it's at least closer to the designs.

chosww commented:
Thanks @the-t-in-rtf for making the updates. I see some other parts we can update to make it closer to the design:

- Each section in Actions panel needs box-shadow at the bottom: in Figma, the details are in Program tab 
- Background-color for class App__command-palette-command-container needs to be updated in Grayscale and Contrast themes
- Each section in Actions panel needs round border around (it already has border-radius set, but need to make it little bit rounder), and in the dark theme, we need darker border colour than the background for the container
- Make the border-bottom of the title (Movements and Controls) a little thicker, 2px would be good
- Loop block color is incorrect in light theme (need to match with the one in design system tab in Figma)

the-t-in-rtf commented:
Thanks, @chosww.  I incorporated your feedback and pushed the changes.

the-t-in-rtf commented:
In the "program" page, Figma uses`#3044E` for the fill colour in light mode, and `#FFFFFF` for the stroke.  What I've done matches that, but I can see that it doesn't match the "design system" page.  I used the hover styles from the "design system" page, which don't have much contrast with the "program" fill colour.  Since the "design system" work seems newer, I've updated to use the "design system" colour for the fill.

the-t-in-rtf commented:
I fixed the contrast issues you found.

I also had a go at fixing the border-related styles (focus, selected) for the loop block.  With the built-in drop shadow I had to increase the size of the outer button so that the SVG would appear roughly centred within the border.

chosww commented:
Thanks @the-t-in-rtf, the fix you made for the focus and selected is a good improvement, and looks good enough to me. I see a new issue with the change you made to apply proper border colour on section heading, there are two borders with different colour on top of each other in the contrast theme now. I suggest removing style you added on App__command-palette-section-heading and override colour of App__command-palette-section-heading-container in the contrast theme. 

Also, for 1.2 release, it would make sense to use #000000 for black as we didn't merge the changes you made before to update #000000 to #1E1E1E. I talked with Simon about that yesterday and they think that sounds reasonable. So let's update the new ActionPanel background in the contrast theme to use #000000.

the-t-in-rtf commented:
> Also, for 1.2 release, it would make sense to use #000000 for black as we didn't merge the changes you made before to update #000000 to #1E1E1E. I talked with Simon about that yesterday and they think that sounds reasonable. So let's update the new ActionPanel background in the contrast theme to use #000000.

I looked for and think I've fixed all of the`#1E1E1E` colours I introduced in this pull.  There are other instances of `#1E1E1E` coming in from other work, I didn't touch those for now.

the-t-in-rtf commented:
> I suggest removing style you added on App__command-palette-section-heading and override colour of App__command-palette-section-heading-container in the contrast theme.

I removed the extra bottom border.  I wasn't sure what you meant by the heading container, if I do that instead of updating the colour of the heading, other default styles trump and the colour isn't set correctly.  Let me know if what I've done is acceptable.

the-t-in-rtf commented:
We just noticed in reviewing this that the simplification menu doesn't correctly detect whether a loop block is already used in the program, so it's possible to deactivate the loop block even if it's already used.  The wording is also not correct, i.e. the "used" wording doesn't appear once a loop block is used.  I suspect this may be down to the way we use "loop" for the UI but "startLoop" and "endLoop" in the program.

the-t-in-rtf commented:
I updated the ProgramSequence code to report that "loop" is used if either "startLoop" or "endLoop" are encountered in the program.

chosww commented:
> > I suggest removing style you added on App__command-palette-section-heading and override colour of App__command-palette-section-heading-container in the contrast theme.
> 
> I removed the extra bottom border. I wasn't sure what you meant by the heading container, if I do that instead of updating the colour of the heading, other default styles trump and the colour isn't set correctly. Let me know if what I've done is acceptable.

Thanks @the-t-in-rtf, the double border issue is resolved now, but the contrast theme's border colour on "App__command-palette-section-heading-container" is still incorrect. We need to add a line in Themes.scss for the contrast theme to apply white border on the class.

the-t-in-rtf commented:
As discussed in chat, I updated to make the bottom border of the section heading container white, as it is in Figma for the contrast theme.

sbates-idrc commented:
The selection of which loop SVG to use (with or without shadow) isn't consistent with the shadow on the other actions.

* In the Actions Panel, the loop has a shadow, but should not
* In the Available Actions dialog, the loop does not have a shadow, but should

Screenshot in the Grayscale theme:

![image](https://user-images.githubusercontent.com/80360967/156609704-9083ac39-55b1-48c9-966a-5866fa06f2a7.png)

sbates-idrc commented:
I'm looking at how the loop SVG selection is made in `CommandBlock` and `ActionsMenuItem`. I think that `CommandBlock` is getting messy and could do with a refactor. But for this PR, I think we should leave things as is. So, I think to address the loop SVG selection, we can keep the current structure but just rename:

* Loop.svg -> LoopWithShadow.svg
* LoopNoShadow.svg -> Loop.svg
* Replace the `['loopNoShadow', LoopNoShadow]` entry in `commandBlockIconTypes` with `['loopWithShadow', LoopWithShadow]`
* Update the logic in `ActionMenuItem` to use `loopWithShadow`

the-t-in-rtf commented on src/types.js:
```diff
@@ -3,6 +3,12 @@ import CharacterState from './CharacterState';
 
+export type DisplayedCommandName =
+    'forward1' | 'backward1' |
+    'left45' | 'left90' | 'left180' |
+    'right45' | 'right90' | 'right180' |
```
Done.

the-t-in-rtf commented on src/KeyboardInputSchemes.js:
```diff
@@ -150,34 +146,18 @@ const ExtendedKeyboardSequences: KeyboardInputScheme = {
+                backward: {
                     keyDef: { key: "1"},
                     actionName: "selectBackward1"
-                },
-                backward2: {
-                    keyDef: { key: "2"},
-                    actionName: "selectBackward2"
-                },
-                backward3: {
-                    keyDef: { key: "3"},
-                    actionName: "selectBackward3"
                 }
             },
```
Done.

the-t-in-rtf commented on src/App.js:
```diff
@@ -935,6 +923,20 @@ export class App extends React.Component<AppProps, AppState> {
         });
     }
 
+    handleToggleAllowedCommand = (event: Event, commandName: DisplayedCommandName) => {
+        // TODO: Use the function form of setState() as the new state
+        //       depends on the current state
+        const currentIsDisallowed = this.state.disallowedActions[commandName];
+        if (currentIsDisallowed) {
+            const newDisallowedActions= Object.assign({}, this.state.disallowedActions);
+            delete newDisallowedActions[commandName];
+            this.setState({ disallowedActions: newDisallowedActions})
+        }
+        else if (this.state.programSequence.usesAction(commandName)) {
+            event.preventDefault();
+        }
+    }
```
Removed.

the-t-in-rtf commented on src/messages.json:
```diff
@@ -82,6 +77,7 @@
         "CommandInfo.previousStep.loop": "before step {previousStepNumber} {command}",
         "CommandInfo.previousStep.startLoop": "out of loop {loopLabel}",
         "CommandInfo.nextStep": "after step {nextStepNumber} {command}",
+        "CommandPalette.controlsTitle": "Controls",
```
Done.

the-t-in-rtf commented:
> So, I think to address the loop SVG selection, we can keep the current structure but just rename:
> 
> * Loop.svg -> LoopWithShadow.svg
> * LoopNoShadow.svg -> Loop.svg
> * Replace the `['loopNoShadow', LoopNoShadow]` entry in `commandBlockIconTypes` with `['loopWithShadow', LoopWithShadow]`
> * Update the logic in `ActionMenuItem` to use `loopWithShadow`

I made the requested changes.  The shadows don't match perfectly in the actions simplification modal, but at least each icon better matches its peers now.
